From: Egbert Eich <eich@suse.de>
Date: Sat May 3 19:06:24 2014 +0200
Subject: drm/mgag200: Cleanup cursor BOs properly
Patch-mainline: Not yet,  upstreaming planned after more testing.
References: bnc#876097 

Cursor BOs should be cleaned up properly on error or when unloading
the driver.

Signed-off-by: Egbert Eich <eich@suse.de>
Signed-off-by: Egbert Eich <eich@suse.com>
---
 drivers/gpu/drm/mgag200/mgag200_main.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)
diff --git a/drivers/gpu/drm/mgag200/mgag200_main.c b/drivers/gpu/drm/mgag200/mgag200_main.c
index 7b12fc2..e3f3d4e 100644
--- a/drivers/gpu/drm/mgag200/mgag200_main.c
+++ b/drivers/gpu/drm/mgag200/mgag200_main.c
@@ -347,9 +347,12 @@ int mgag200_driver_load(struct drm_device *dev, unsigned long flags)
 	/* Make small buffers to store a hardware cursor (double buffered icon updates) */
 	mgag200_bo_create(dev, roundup(48*64, PAGE_SIZE), 0, 0,
 					  &mdev->cursor.pixels_1);
-	mgag200_bo_create(dev, roundup(48*64, PAGE_SIZE), 0, 0,
-					  &mdev->cursor.pixels_2);
+	if (mdev->cursor.pixels_1)
+		mgag200_bo_create(dev, roundup(48*64, PAGE_SIZE), 0, 0,
+				                  &mdev->cursor.pixels_2);
 	if (!mdev->cursor.pixels_2 || !mdev->cursor.pixels_1) {
+		if (mdev->cursor.pixels_1)
+			drm_gem_object_unreference_unlocked(&mdev->cursor.pixels_1->gem);
 		mdev->cursor.pixels_1 = NULL;
 		mdev->cursor.pixels_2 = NULL;
 		dev_warn(&dev->pdev->dev,
@@ -377,6 +380,10 @@ int mgag200_driver_unload(struct drm_device *dev)
 	if (mdev == NULL)
 		return 0;
 	mgag200_modeset_fini(mdev);
+	if (mdev->cursor.pixels_1)
+		drm_gem_object_unreference_unlocked(&mdev->cursor.pixels_1->gem);
+	if (mdev->cursor.pixels_2)
+		drm_gem_object_unreference_unlocked(&mdev->cursor.pixels_2->gem);
 	mgag200_fbdev_fini(mdev);
 	drm_mode_config_cleanup(dev);
 	mgag200_mm_fini(mdev);
